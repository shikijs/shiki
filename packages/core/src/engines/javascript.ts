import { onigurumaToRegexp } from 'oniguruma-to-js'
import type { JavaScriptRegexEngineOptions, PatternScanner, RegexEngine, RegexEngineString } from '../types'

const MAX = 4294967295

/**
 * The default RegExp constructor for JavaScript regex engine.
 */
export function defaultJavaScriptRegexConstructor(pattern: string): RegExp {
  return onigurumaToRegexp(
    pattern
      // YAML specific handling; TODO: move to tm-grammars
      .replaceAll('[^\\s[-?:,\\[\\]{}#&*!|>\'"%@`]]', '[^\\s\\-?:,\\[\\]{}#&*!|>\'"%@`]'),
    {
      flags: 'dgm',
      ignoreContiguousAnchors: true,
    },
  )
}

export class JavaScriptScanner implements PatternScanner {
  regexps: (RegExp | null)[]

  constructor(
    public patterns: string[],
    public cache: Map<string, RegExp | Error>,
    public forgiving: boolean,
    public regexConstructor: (pattern: string) => RegExp = defaultJavaScriptRegexConstructor,
  ) {
    this.regexps = patterns.map((p) => {
      const cached = cache?.get(p)
      if (cached) {
        if (cached instanceof RegExp) {
          return cached
        }
        if (forgiving)
          return null
        throw cached
      }
      try {
        const regex = regexConstructor(p)
        cache?.set(p, regex)
        return regex
      }
      catch (e) {
        cache?.set(p, e as Error)
        if (forgiving)
          return null
        // console.error({ ...e })
        throw e
      }
    })
  }

  findNextMatchSync(string: string | RegexEngineString, startPosition: number) {
    const str = typeof string === 'string'
      ? string
      : string.content
    const pending: [index: number, match: RegExpExecArray][] = []

    function toResult(index: number, match: RegExpExecArray) {
      return {
        index,
        captureIndices: match.indices!.map((indice) => {
          if (indice == null) {
            return {
              end: MAX,
              start: MAX,
              length: 0,
            }
          }
          return {
            start: indice[0],
            length: indice[1] - indice[0],
            end: indice[1],
          }
        }),
      }
    }

    for (let i = 0; i < this.regexps.length; i++) {
      const regexp = this.regexps[i]
      if (!regexp)
        continue
      try {
        regexp.lastIndex = startPosition
        const match = regexp.exec(str)
        if (!match)
          continue
        // If the match is at the start position, return it immediately
        if (match.index === startPosition) {
          return toResult(i, match)
        }
        // Otherwise, store it for later
        pending.push([i, match])
      }
      catch (e) {
        if (this.forgiving)
          continue
        throw e
      }
    }

    // Find the closest match to the start position
    if (pending.length) {
      const minIndex = Math.min(...pending.map(m => m[1].index))
      for (const [i, match] of pending) {
        if (match.index === minIndex) {
          return toResult(i, match)
        }
      }
    }

    return null
  }
}

/**
 * Use the modern JavaScript RegExp engine to implement the OnigScanner.
 *
 * As Oniguruma regex is more powerful than JavaScript regex, some patterns may not be supported.
 * Errors will be thrown when parsing TextMate grammars with unsupported patterns.
 * Set `forgiving` to `true` to ignore these errors and skip the unsupported patterns.
 *
 * @experimental
 */
export function createJavaScriptRegexEngine(options: JavaScriptRegexEngineOptions = {}): RegexEngine {
  const {
    forgiving = false,
    cache = new Map(),
  } = options

  return {
    createScanner(patterns: string[]) {
      return new JavaScriptScanner(patterns, cache, forgiving, options.regexConstructor)
    },
    createString(s: string) {
      return {
        content: s,
      }
    },
  }
}
