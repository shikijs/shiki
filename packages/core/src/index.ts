// Types
export * from './constructors/bundle-factory'

// Constructors
export * from './constructors/highlighter'
export { createShikiInternal, getShikiInternal, setDefaultWasmLoader } from './constructors/internal'
export { createShikiInternalSync } from './constructors/internal-sync'
export { createJavaScriptRegexEngine, defaultJavaScriptRegexConstructor } from './engines/javascript'

// Engines
export { createWasmOnigEngine, loadWasm } from './engines/wasm'
export { ShikiError } from './error'

export { codeToHast, tokensToHast } from './highlight/code-to-hast'

export { codeToHtml, hastToHtml } from './highlight/code-to-html'
export { codeToTokens } from './highlight/code-to-tokens'
export { tokenizeAnsiWithTheme } from './highlight/code-to-tokens-ansi'
// Low-level Highlighting
export { codeToTokensBase, tokenizeWithTheme } from './highlight/code-to-tokens-base'
export { codeToTokensWithThemes } from './highlight/code-to-tokens-themes'
// TextMate Utilities
export { normalizeTheme } from './textmate/normalize-theme'

export { transformerDecorations } from './transformer-decorations'
export * from './types'
// Utils and Misc
export * from './utils'

// Deprecated
export {
  /**
   * @deprecated Use `EncodedTokenMetadata` from `@shikijs/vscode-textmate` instead.
   */
  EncodedTokenMetadata as StackElementMetadata,
} from '@shikijs/vscode-textmate'
