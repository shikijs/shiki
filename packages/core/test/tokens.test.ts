import { createJavaScriptRegexEngine } from 'shiki'
import { describe, expect, it } from 'vitest'
import { codeToTokens, codeToTokensBase, createShikiInternal } from '../src'

it('includeExplanation', async () => {
  using engine = await createShikiInternal({
    themes: [
      import('@shikijs/themes/vitesse-dark'),
    ],
    langs: [
      import('@shikijs/langs/javascript'),
    ],
    engine: createJavaScriptRegexEngine(),
  })

  const code = 'console.log("hello")'
  const caseFalse = await codeToTokensBase(engine, code, { lang: 'js', theme: 'vitesse-dark', includeExplanation: false })
  const caseTrue = await codeToTokensBase(engine, code, { lang: 'js', theme: 'vitesse-dark', includeExplanation: true })
  const caseScopeName = await codeToTokensBase(engine, code, { lang: 'js', theme: 'vitesse-dark', includeExplanation: 'scopeName' })

  expect(caseFalse).toMatchSnapshot('false')
  expect(caseTrue).toMatchSnapshot('true')
  expect(caseScopeName).toMatchSnapshot('scopeName')
})

describe('defaultColor light-dark()', () => {
  it('basic', async () => {
    using engine = await createShikiInternal({
      themes: [
        import('@shikijs/themes/vitesse-light'),
        import('@shikijs/themes/vitesse-dark'),
      ],
      langs: [
        import('@shikijs/langs/javascript'),
      ],
      engine: createJavaScriptRegexEngine(),
    })

    const code = 'console.log("hello")'
    const caseCssVars = codeToTokens(engine, code, {
      lang: 'js',
      themes: { light: 'vitesse-light', dark: 'vitesse-dark' },
      defaultColor: false,
    })
    const caseLightDark = codeToTokens(engine, code, {
      lang: 'js',
      themes: { light: 'vitesse-light', dark: 'vitesse-dark' },
      defaultColor: 'light-dark()',
    })

    expect(caseCssVars).toMatchSnapshot('css-vars')
    expect(caseLightDark).toMatchSnapshot('light-dark()')
  })

  it('defaultColor light-dark() with multiple themes', async () => {
    using engine = await createShikiInternal({
      themes: [
        import('@shikijs/themes/vitesse-light'),
        import('@shikijs/themes/vitesse-dark'),
        import('@shikijs/themes/github-dark'),
      ],
      langs: [
        import('@shikijs/langs/javascript'),
      ],
      engine: createJavaScriptRegexEngine(),
    })

    const code = 'console.log("hello")'
    const caseCssVars = codeToTokens(engine, code, {
      lang: 'js',
      themes: {
        dark: 'vitesse-dark',
        light: 'vitesse-light',
        custom: 'github-dark',
      },
      defaultColor: false,
    })
    const caseLightDark = codeToTokens(engine, code, {
      lang: 'js',
      themes: {
        dark: 'vitesse-dark',
        light: 'vitesse-light',
        custom: 'github-dark',
      },
      defaultColor: 'light-dark()',
    })

    expect(caseCssVars).toMatchSnapshot('css-vars')
    expect(caseLightDark).toMatchSnapshot('light-dark()')
  })

  it('should throw when no light or dark theme is provided', async () => {
    using engine = await createShikiInternal({
      themes: [
        import('@shikijs/themes/vitesse-light'),
        import('@shikijs/themes/vitesse-dark'),
        import('@shikijs/themes/github-dark'),
      ],
      langs: [
        import('@shikijs/langs/javascript'),
      ],
      engine: createJavaScriptRegexEngine(),
    })

    const code = 'console.log("hello")'

    expect(() => {
      codeToTokens(engine, code, {
        lang: 'js',
        themes: {
          dark2: 'vitesse-dark',
          light1: 'vitesse-light',
        },
        defaultColor: 'light-dark()',
      })
    })
      .toThrowErrorMatchingInlineSnapshot(`[ShikiError: When using \`defaultColor: "light-dark()"\`, you must provide both \`light\` and \`dark\` themes]`)

    expect(() => {
      codeToTokens(engine, code, {
        lang: 'js',
        themes: {
          dark: 'vitesse-dark',
          light1: 'vitesse-light',
        },
        defaultColor: 'light-dark()',
      })
    })
      .toThrowErrorMatchingInlineSnapshot(`[ShikiError: When using \`defaultColor: "light-dark()"\`, you must provide both \`light\` and \`dark\` themes]`)

    // not throw when only one theme is provided
    codeToTokens(engine, code, {
      lang: 'js',
      theme: 'vitesse-dark',
      defaultColor: 'light-dark()',
    })
  })
})
