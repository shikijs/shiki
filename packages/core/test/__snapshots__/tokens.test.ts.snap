// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`defaultColor light-dark() > basic > css-vars 1`] = `
{
  "bg": "--shiki-light-bg:#ffffff;--shiki-dark-bg:#121212",
  "fg": "--shiki-light:#393a34;--shiki-dark:#dbd7caee",
  "grammarState": {
    "lang": "javascript",
    "scopes": [
      "source.js",
    ],
    "theme": "vitesse-light",
    "themes": [
      "vitesse-light",
      "vitesse-dark",
    ],
  },
  "rootStyle": "--shiki-light:#393a34;--shiki-dark:#dbd7caee;--shiki-light-bg:#ffffff;--shiki-dark-bg:#121212",
  "themeName": "shiki-themes vitesse-light vitesse-dark",
  "tokens": [
    [
      {
        "content": "console",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-dark": "#BD976A",
          "--shiki-light": "#B07D48",
        },
        "offset": 0,
      },
      {
        "content": ".",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-dark": "#666666",
          "--shiki-light": "#999999",
        },
        "offset": 7,
      },
      {
        "content": "log",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-dark": "#80A665",
          "--shiki-light": "#59873A",
        },
        "offset": 8,
      },
      {
        "content": "(",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-dark": "#666666",
          "--shiki-light": "#999999",
        },
        "offset": 11,
      },
      {
        "content": """,
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-dark": "#C98A7D77",
          "--shiki-light": "#B5695977",
        },
        "offset": 12,
      },
      {
        "content": "hello",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-dark": "#C98A7D",
          "--shiki-light": "#B56959",
        },
        "offset": 13,
      },
      {
        "content": """,
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-dark": "#C98A7D77",
          "--shiki-light": "#B5695977",
        },
        "offset": 18,
      },
      {
        "content": ")",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-dark": "#666666",
          "--shiki-light": "#999999",
        },
        "offset": 19,
      },
    ],
  ],
}
`;

exports[`defaultColor light-dark() > basic > light-dark() 1`] = `
{
  "bg": "light-dark(#ffffff, #121212);--shiki-light-bg:#ffffff;--shiki-dark-bg:#121212",
  "fg": "light-dark(#393a34, #dbd7caee);--shiki-light:#393a34;--shiki-dark:#dbd7caee",
  "grammarState": {
    "lang": "javascript",
    "scopes": [
      "source.js",
    ],
    "theme": "vitesse-light",
    "themes": [
      "vitesse-light",
      "vitesse-dark",
    ],
  },
  "rootStyle": undefined,
  "themeName": "shiki-themes vitesse-light vitesse-dark",
  "tokens": [
    [
      {
        "content": "console",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-dark": "#BD976A",
          "--shiki-light": "#B07D48",
          "color": "light-dark(#B07D48, #BD976A)",
        },
        "offset": 0,
      },
      {
        "content": ".",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-dark": "#666666",
          "--shiki-light": "#999999",
          "color": "light-dark(#999999, #666666)",
        },
        "offset": 7,
      },
      {
        "content": "log",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-dark": "#80A665",
          "--shiki-light": "#59873A",
          "color": "light-dark(#59873A, #80A665)",
        },
        "offset": 8,
      },
      {
        "content": "(",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-dark": "#666666",
          "--shiki-light": "#999999",
          "color": "light-dark(#999999, #666666)",
        },
        "offset": 11,
      },
      {
        "content": """,
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-dark": "#C98A7D77",
          "--shiki-light": "#B5695977",
          "color": "light-dark(#B5695977, #C98A7D77)",
        },
        "offset": 12,
      },
      {
        "content": "hello",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-dark": "#C98A7D",
          "--shiki-light": "#B56959",
          "color": "light-dark(#B56959, #C98A7D)",
        },
        "offset": 13,
      },
      {
        "content": """,
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-dark": "#C98A7D77",
          "--shiki-light": "#B5695977",
          "color": "light-dark(#B5695977, #C98A7D77)",
        },
        "offset": 18,
      },
      {
        "content": ")",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-dark": "#666666",
          "--shiki-light": "#999999",
          "color": "light-dark(#999999, #666666)",
        },
        "offset": 19,
      },
    ],
  ],
}
`;

exports[`defaultColor light-dark() > defaultColor light-dark() with multiple themes > css-vars 1`] = `
{
  "bg": "--shiki-dark-bg:#121212;--shiki-light-bg:#ffffff;--shiki-custom-bg:#24292e",
  "fg": "--shiki-dark:#dbd7caee;--shiki-light:#393a34;--shiki-custom:#e1e4e8",
  "grammarState": {
    "lang": "javascript",
    "scopes": [
      "source.js",
    ],
    "theme": "vitesse-dark",
    "themes": [
      "vitesse-dark",
      "vitesse-light",
      "github-dark",
    ],
  },
  "rootStyle": "--shiki-dark:#dbd7caee;--shiki-light:#393a34;--shiki-custom:#e1e4e8;--shiki-dark-bg:#121212;--shiki-light-bg:#ffffff;--shiki-custom-bg:#24292e",
  "themeName": "shiki-themes vitesse-dark vitesse-light github-dark",
  "tokens": [
    [
      {
        "content": "console",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-custom": "#E1E4E8",
          "--shiki-dark": "#BD976A",
          "--shiki-light": "#B07D48",
        },
        "offset": 0,
      },
      {
        "content": ".",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-custom": "#E1E4E8",
          "--shiki-dark": "#666666",
          "--shiki-light": "#999999",
        },
        "offset": 7,
      },
      {
        "content": "log",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-custom": "#B392F0",
          "--shiki-dark": "#80A665",
          "--shiki-light": "#59873A",
        },
        "offset": 8,
      },
      {
        "content": "(",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-custom": "#E1E4E8",
          "--shiki-dark": "#666666",
          "--shiki-light": "#999999",
        },
        "offset": 11,
      },
      {
        "content": """,
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-custom": "#9ECBFF",
          "--shiki-dark": "#C98A7D77",
          "--shiki-light": "#B5695977",
        },
        "offset": 12,
      },
      {
        "content": "hello",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-custom": "#9ECBFF",
          "--shiki-dark": "#C98A7D",
          "--shiki-light": "#B56959",
        },
        "offset": 13,
      },
      {
        "content": """,
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-custom": "#9ECBFF",
          "--shiki-dark": "#C98A7D77",
          "--shiki-light": "#B5695977",
        },
        "offset": 18,
      },
      {
        "content": ")",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-custom": "#E1E4E8",
          "--shiki-dark": "#666666",
          "--shiki-light": "#999999",
        },
        "offset": 19,
      },
    ],
  ],
}
`;

exports[`defaultColor light-dark() > defaultColor light-dark() with multiple themes > light-dark() 1`] = `
{
  "bg": "light-dark(#ffffff, #121212);--shiki-dark-bg:#121212;--shiki-light-bg:#ffffff;--shiki-custom-bg:#24292e",
  "fg": "light-dark(#393a34, #dbd7caee);--shiki-dark:#dbd7caee;--shiki-light:#393a34;--shiki-custom:#e1e4e8",
  "grammarState": {
    "lang": "javascript",
    "scopes": [
      "source.js",
    ],
    "theme": "vitesse-dark",
    "themes": [
      "vitesse-dark",
      "vitesse-light",
      "github-dark",
    ],
  },
  "rootStyle": undefined,
  "themeName": "shiki-themes vitesse-dark vitesse-light github-dark",
  "tokens": [
    [
      {
        "content": "console",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-custom": "#E1E4E8",
          "--shiki-dark": "#BD976A",
          "--shiki-light": "#B07D48",
          "color": "light-dark(#B07D48, #BD976A)",
        },
        "offset": 0,
      },
      {
        "content": ".",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-custom": "#E1E4E8",
          "--shiki-dark": "#666666",
          "--shiki-light": "#999999",
          "color": "light-dark(#999999, #666666)",
        },
        "offset": 7,
      },
      {
        "content": "log",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-custom": "#B392F0",
          "--shiki-dark": "#80A665",
          "--shiki-light": "#59873A",
          "color": "light-dark(#59873A, #80A665)",
        },
        "offset": 8,
      },
      {
        "content": "(",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-custom": "#E1E4E8",
          "--shiki-dark": "#666666",
          "--shiki-light": "#999999",
          "color": "light-dark(#999999, #666666)",
        },
        "offset": 11,
      },
      {
        "content": """,
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-custom": "#9ECBFF",
          "--shiki-dark": "#C98A7D77",
          "--shiki-light": "#B5695977",
          "color": "light-dark(#B5695977, #C98A7D77)",
        },
        "offset": 12,
      },
      {
        "content": "hello",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-custom": "#9ECBFF",
          "--shiki-dark": "#C98A7D",
          "--shiki-light": "#B56959",
          "color": "light-dark(#B56959, #C98A7D)",
        },
        "offset": 13,
      },
      {
        "content": """,
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-custom": "#9ECBFF",
          "--shiki-dark": "#C98A7D77",
          "--shiki-light": "#B5695977",
          "color": "light-dark(#B5695977, #C98A7D77)",
        },
        "offset": 18,
      },
      {
        "content": ")",
        "explanation": undefined,
        "htmlStyle": {
          "--shiki-custom": "#E1E4E8",
          "--shiki-dark": "#666666",
          "--shiki-light": "#999999",
          "color": "light-dark(#999999, #666666)",
        },
        "offset": 19,
      },
    ],
  ],
}
`;

exports[`includeExplanation > false 1`] = `
[
  [
    {
      "color": "#BD976A",
      "content": "console",
      "fontStyle": 0,
      "offset": 0,
    },
    {
      "color": "#666666",
      "content": ".",
      "fontStyle": 0,
      "offset": 7,
    },
    {
      "color": "#80A665",
      "content": "log",
      "fontStyle": 0,
      "offset": 8,
    },
    {
      "color": "#666666",
      "content": "(",
      "fontStyle": 0,
      "offset": 11,
    },
    {
      "color": "#C98A7D77",
      "content": """,
      "fontStyle": 0,
      "offset": 12,
    },
    {
      "color": "#C98A7D",
      "content": "hello",
      "fontStyle": 0,
      "offset": 13,
    },
    {
      "color": "#C98A7D77",
      "content": """,
      "fontStyle": 0,
      "offset": 18,
    },
    {
      "color": "#666666",
      "content": ")",
      "fontStyle": 0,
      "offset": 19,
    },
  ],
]
`;

exports[`includeExplanation > scopeName 1`] = `
[
  [
    {
      "color": "#BD976A",
      "content": "console",
      "explanation": [
        {
          "content": "console",
          "scopes": [
            {
              "scopeName": "source.js",
            },
            {
              "scopeName": "meta.function-call.js",
            },
            {
              "scopeName": "variable.other.object.js",
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 0,
    },
    {
      "color": "#666666",
      "content": ".",
      "explanation": [
        {
          "content": ".",
          "scopes": [
            {
              "scopeName": "source.js",
            },
            {
              "scopeName": "meta.function-call.js",
            },
            {
              "scopeName": "punctuation.accessor.js",
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 7,
    },
    {
      "color": "#80A665",
      "content": "log",
      "explanation": [
        {
          "content": "log",
          "scopes": [
            {
              "scopeName": "source.js",
            },
            {
              "scopeName": "meta.function-call.js",
            },
            {
              "scopeName": "entity.name.function.js",
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 8,
    },
    {
      "color": "#666666",
      "content": "(",
      "explanation": [
        {
          "content": "(",
          "scopes": [
            {
              "scopeName": "source.js",
            },
            {
              "scopeName": "meta.brace.round.js",
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 11,
    },
    {
      "color": "#C98A7D77",
      "content": """,
      "explanation": [
        {
          "content": """,
          "scopes": [
            {
              "scopeName": "source.js",
            },
            {
              "scopeName": "string.quoted.double.js",
            },
            {
              "scopeName": "punctuation.definition.string.begin.js",
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 12,
    },
    {
      "color": "#C98A7D",
      "content": "hello",
      "explanation": [
        {
          "content": "hello",
          "scopes": [
            {
              "scopeName": "source.js",
            },
            {
              "scopeName": "string.quoted.double.js",
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 13,
    },
    {
      "color": "#C98A7D77",
      "content": """,
      "explanation": [
        {
          "content": """,
          "scopes": [
            {
              "scopeName": "source.js",
            },
            {
              "scopeName": "string.quoted.double.js",
            },
            {
              "scopeName": "punctuation.definition.string.end.js",
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 18,
    },
    {
      "color": "#666666",
      "content": ")",
      "explanation": [
        {
          "content": ")",
          "scopes": [
            {
              "scopeName": "source.js",
            },
            {
              "scopeName": "meta.brace.round.js",
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 19,
    },
  ],
]
`;

exports[`includeExplanation > true 1`] = `
[
  [
    {
      "color": "#BD976A",
      "content": "console",
      "explanation": [
        {
          "content": "console",
          "scopes": [
            {
              "scopeName": "source.js",
              "themeMatches": [],
            },
            {
              "scopeName": "meta.function-call.js",
              "themeMatches": [],
            },
            {
              "scopeName": "variable.other.object.js",
              "themeMatches": [
                {
                  "scope": [
                    "variable",
                    "identifier",
                  ],
                  "settings": {
                    "foreground": "#bd976a",
                  },
                },
              ],
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 0,
    },
    {
      "color": "#666666",
      "content": ".",
      "explanation": [
        {
          "content": ".",
          "scopes": [
            {
              "scopeName": "source.js",
              "themeMatches": [],
            },
            {
              "scopeName": "meta.function-call.js",
              "themeMatches": [],
            },
            {
              "scopeName": "punctuation.accessor.js",
              "themeMatches": [
                {
                  "scope": [
                    "delimiter.bracket",
                    "delimiter",
                    "invalid.illegal.character-not-allowed-here.html",
                    "keyword.operator.rest",
                    "keyword.operator.spread",
                    "keyword.operator.type.annotation",
                    "keyword.operator.relational",
                    "keyword.operator.assignment",
                    "keyword.operator.type",
                    "meta.brace",
                    "meta.tag.block.any.html",
                    "meta.tag.inline.any.html",
                    "meta.tag.structure.input.void.html",
                    "meta.type.annotation",
                    "meta.embedded.block.github-actions-expression",
                    "storage.type.function.arrow",
                    "meta.objectliteral.ts",
                    "punctuation",
                    "punctuation.definition.string.begin.html.vue",
                    "punctuation.definition.string.end.html.vue",
                  ],
                  "settings": {
                    "foreground": "#666666",
                  },
                },
              ],
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 7,
    },
    {
      "color": "#80A665",
      "content": "log",
      "explanation": [
        {
          "content": "log",
          "scopes": [
            {
              "scopeName": "source.js",
              "themeMatches": [],
            },
            {
              "scopeName": "meta.function-call.js",
              "themeMatches": [],
            },
            {
              "scopeName": "entity.name.function.js",
              "themeMatches": [
                {
                  "scope": [
                    "entity",
                    "entity.name",
                  ],
                  "settings": {
                    "foreground": "#80a665",
                  },
                },
                {
                  "scope": "entity.name.function",
                  "settings": {
                    "foreground": "#80a665",
                  },
                },
              ],
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 8,
    },
    {
      "color": "#666666",
      "content": "(",
      "explanation": [
        {
          "content": "(",
          "scopes": [
            {
              "scopeName": "source.js",
              "themeMatches": [],
            },
            {
              "scopeName": "meta.brace.round.js",
              "themeMatches": [
                {
                  "scope": [
                    "delimiter.bracket",
                    "delimiter",
                    "invalid.illegal.character-not-allowed-here.html",
                    "keyword.operator.rest",
                    "keyword.operator.spread",
                    "keyword.operator.type.annotation",
                    "keyword.operator.relational",
                    "keyword.operator.assignment",
                    "keyword.operator.type",
                    "meta.brace",
                    "meta.tag.block.any.html",
                    "meta.tag.inline.any.html",
                    "meta.tag.structure.input.void.html",
                    "meta.type.annotation",
                    "meta.embedded.block.github-actions-expression",
                    "storage.type.function.arrow",
                    "meta.objectliteral.ts",
                    "punctuation",
                    "punctuation.definition.string.begin.html.vue",
                    "punctuation.definition.string.end.html.vue",
                  ],
                  "settings": {
                    "foreground": "#666666",
                  },
                },
              ],
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 11,
    },
    {
      "color": "#C98A7D77",
      "content": """,
      "explanation": [
        {
          "content": """,
          "scopes": [
            {
              "scopeName": "source.js",
              "themeMatches": [],
            },
            {
              "scopeName": "string.quoted.double.js",
              "themeMatches": [
                {
                  "scope": [
                    "string",
                    "string punctuation.section.embedded source",
                    "attribute.value",
                  ],
                  "settings": {
                    "foreground": "#c98a7d",
                  },
                },
              ],
            },
            {
              "scopeName": "punctuation.definition.string.begin.js",
              "themeMatches": [
                {
                  "scope": [
                    "delimiter.bracket",
                    "delimiter",
                    "invalid.illegal.character-not-allowed-here.html",
                    "keyword.operator.rest",
                    "keyword.operator.spread",
                    "keyword.operator.type.annotation",
                    "keyword.operator.relational",
                    "keyword.operator.assignment",
                    "keyword.operator.type",
                    "meta.brace",
                    "meta.tag.block.any.html",
                    "meta.tag.inline.any.html",
                    "meta.tag.structure.input.void.html",
                    "meta.type.annotation",
                    "meta.embedded.block.github-actions-expression",
                    "storage.type.function.arrow",
                    "meta.objectliteral.ts",
                    "punctuation",
                    "punctuation.definition.string.begin.html.vue",
                    "punctuation.definition.string.end.html.vue",
                  ],
                  "settings": {
                    "foreground": "#666666",
                  },
                },
                {
                  "scope": [
                    "punctuation.definition.string",
                  ],
                  "settings": {
                    "foreground": "#c98a7d77",
                  },
                },
              ],
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 12,
    },
    {
      "color": "#C98A7D",
      "content": "hello",
      "explanation": [
        {
          "content": "hello",
          "scopes": [
            {
              "scopeName": "source.js",
              "themeMatches": [],
            },
            {
              "scopeName": "string.quoted.double.js",
              "themeMatches": [
                {
                  "scope": [
                    "string",
                    "string punctuation.section.embedded source",
                    "attribute.value",
                  ],
                  "settings": {
                    "foreground": "#c98a7d",
                  },
                },
              ],
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 13,
    },
    {
      "color": "#C98A7D77",
      "content": """,
      "explanation": [
        {
          "content": """,
          "scopes": [
            {
              "scopeName": "source.js",
              "themeMatches": [],
            },
            {
              "scopeName": "string.quoted.double.js",
              "themeMatches": [
                {
                  "scope": [
                    "string",
                    "string punctuation.section.embedded source",
                    "attribute.value",
                  ],
                  "settings": {
                    "foreground": "#c98a7d",
                  },
                },
              ],
            },
            {
              "scopeName": "punctuation.definition.string.end.js",
              "themeMatches": [
                {
                  "scope": [
                    "delimiter.bracket",
                    "delimiter",
                    "invalid.illegal.character-not-allowed-here.html",
                    "keyword.operator.rest",
                    "keyword.operator.spread",
                    "keyword.operator.type.annotation",
                    "keyword.operator.relational",
                    "keyword.operator.assignment",
                    "keyword.operator.type",
                    "meta.brace",
                    "meta.tag.block.any.html",
                    "meta.tag.inline.any.html",
                    "meta.tag.structure.input.void.html",
                    "meta.type.annotation",
                    "meta.embedded.block.github-actions-expression",
                    "storage.type.function.arrow",
                    "meta.objectliteral.ts",
                    "punctuation",
                    "punctuation.definition.string.begin.html.vue",
                    "punctuation.definition.string.end.html.vue",
                  ],
                  "settings": {
                    "foreground": "#666666",
                  },
                },
                {
                  "scope": [
                    "punctuation.definition.string",
                  ],
                  "settings": {
                    "foreground": "#c98a7d77",
                  },
                },
              ],
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 18,
    },
    {
      "color": "#666666",
      "content": ")",
      "explanation": [
        {
          "content": ")",
          "scopes": [
            {
              "scopeName": "source.js",
              "themeMatches": [],
            },
            {
              "scopeName": "meta.brace.round.js",
              "themeMatches": [
                {
                  "scope": [
                    "delimiter.bracket",
                    "delimiter",
                    "invalid.illegal.character-not-allowed-here.html",
                    "keyword.operator.rest",
                    "keyword.operator.spread",
                    "keyword.operator.type.annotation",
                    "keyword.operator.relational",
                    "keyword.operator.assignment",
                    "keyword.operator.type",
                    "meta.brace",
                    "meta.tag.block.any.html",
                    "meta.tag.inline.any.html",
                    "meta.tag.structure.input.void.html",
                    "meta.type.annotation",
                    "meta.embedded.block.github-actions-expression",
                    "storage.type.function.arrow",
                    "meta.objectliteral.ts",
                    "punctuation",
                    "punctuation.definition.string.begin.html.vue",
                    "punctuation.definition.string.end.html.vue",
                  ],
                  "settings": {
                    "foreground": "#666666",
                  },
                },
              ],
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 19,
    },
  ],
]
`;
